common/lib/transport/connectionmanager.js

```diff
@@ -809,70 +785,217 @@ var ConnectionManager = (function() {
+			});
+		};
+
+		Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.startConnect()', 'starting connection');
 		this.startSuspendTimer();
 		this.startTransitionTimer(this.states.connecting);
 
-		var self = this;
-		var auth = this.realtime.auth;
-		var connectErr = function(err) {
-			err = ErrorInfo.fromValues(err);
-			Logger.logAction(Logger.LOG_ERROR, 'ConnectionManager.connectImpl()', 'Connection attempt failed with error; err = ' + err.toString());
-			var state = self.state;
-			if(state == self.states.closing || state == self.states.closed || state == self.states.failed) {
-				/* do nothing */
-				return;
+		if(auth.method == 'basic') {
+			connect();
+		} else {
+			var authOptions = (this.errorReason && Auth.isTokenErr(this.errorReason)) ? {force: true} : null;
+			auth.authorise(null, authOptions, function(err) {
+				if(err) {
+					self.notifyState({state: 'failed', error: err});
+				} else {
+					connect();
+				}
+			});
+		}
+	};
+
+
+	/**
+	 * There are three stages in connecting:
+	 * - preference: if there is a cached transport preference, we try to
+	 *   connect on that. If that fails we remove the preference and fall back
+	 *   to base.
+	 * - base: we try to connect with the best transport that we think will
+	 *   never fail for this browser (usually this is xhr_polling; for very old
+	 *   browsers will be jsonp, for node will be comet). If it doesn't work, we
+	 *   try fallback hosts.
+	 * - upgrade: given a connected transport, we see if there are any better
+	 *   ones, and try to upgrade to them.
+	 *
+	 * connectImpl works out what stage you're at (which is purely a function of
+	 * the current connection state and whether there are any stored preferences),
+	 * and dispatches accordingly. After a transport has been set pending,
+	 * tryATransport calls connectImpl to see if there's another stage to be done.
+	 * */
+	ConnectionManager.prototype.connectImpl = function(transportParams) {
+		var state = this.state.state;
+
+		if(state !== this.states.connecting.state && state !== this.states.connected.state) {
+			/* Only keep trying as long as in the 'connecting' state (or 'connected'
+			 * for upgrading). Any operation can put us into 'disconnected' to cancel
+			 * connection attempts and wait before retrying, or 'failed' to fail. */
+			Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.connectImpl()', 'Must be in connecting state to connect (or connected to upgrade), but was ' + state);
+		} else if(this.pendingTransports.length) {
+			Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.connectImpl()', 'Transports ' + this.pendingTransports[0].toString() + ' currently pending; taking no action');
+		} else if(state == this.states.connected.state) {
+			this.upgradeIfNeeded(transportParams);
+		} else if(this.transports.length > 1 && getTransportPreference()) {
+			this.connectPreference(transportParams);
+		} else {
+			this.connectBase(transportParams);
+		}
+	};
+
```

Comment:
I left an extra line between all the new methods I added, just to make it a bit clearer. Happy to remove if you prefer